### PR TITLE
1D stroke width varying

### PR DESCRIPTION
**Pull request by vcloarec**
strokeWidthVarying

### CODE REVIEW

File: src/app/mesh/qgsmeshstrokewidthvaryingwidget.cpp [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+  whileBlocking( mIgnoreOutOfRangecheckBox )->setChecked( strokeWidthVarying.ignoreOutOfRange() );
+}
+
+void QgsMeshStrokeWidthVaryingButton::setDefaultMinMaxValue( double minimum, double maximum )
+{
+  mMinimumDefaultValue = minimum;
+  mMaximumDefaultValue = maximum;
+}
+
+QgsMeshStrokeWidth QgsMeshStrokeWidthVaryingWidget::varyingStrokeWidth() const
```
Comment (PeterPetrik) [inline src/app/mesh/qgsmeshstrokewidthvaryingwidget.cpp]: strokeWidthSettings

File: src/app/mesh/qgsmeshstrokewidthvaryingwidget.cpp [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+  connect( mValueMaximumLineEdit, &QLineEdit::textEdited, this, &QgsMeshStrokeWidthVaryingWidget::widgetChanged );
+  connect( mWidthMinimumSpinBox, qgis::overload<double>::of( &QgsDoubleSpinBox::valueChanged ),
+           this, &QgsMeshStrokeWidthVaryingWidget::widgetChanged );
+  connect( mWidthMaximumSpinBox, qgis::overload<double>::of( &QgsDoubleSpinBox::valueChanged ),
+           this, &QgsMeshStrokeWidthVaryingWidget::widgetChanged );
+  connect( mIgnoreOutOfRangecheckBox, &QCheckBox::toggled,
+           this, &QgsMeshStrokeWidthVaryingWidget::widgetChanged );
+}
+
+void QgsMeshStrokeWidthVaryingWidget::setVaryingStrokeWidth( const QgsMeshStrokeWidth &strokeWidthVarying )
```
Comment (PeterPetrik) [inline src/app/mesh/qgsmeshstrokewidthvaryingwidget.cpp]: setStrokeWidthSettings

File: src/app/mesh/qgsmeshstrokewidthvaryingwidget.h [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+#include "qgis_app.h"
+#include "ui_qgsmeshvaryingstrokewidthwidgetbase.h"
+#include "qgspanelwidget.h"
+#include "qgsmeshlayerrenderer.h"
+
+
+/**
+ * A widget push button that store stoke width varying and call a widget to set parameters
+ */
+class APP_NO_EXPORT QgsMeshStrokeWidthVaryingButton: public QPushButton
```
Comment (PeterPetrik) [inline src/app/mesh/qgsmeshstrokewidthvaryingwidget.h]: StrokeWidthVarying -> VariableStrokeWidth

File: src/core/mesh/qgsmeshlayer.cpp [hunk 4]
```diff
@@ -488,6 +488,12 @@ void QgsMeshLayer::assignDefaultStyleToDatasetGroup( int groupIndex )
   QgsMeshRendererScalarSettings scalarSettings;
   scalarSettings.setClassificationMinimumMaximum( groupMin, groupMax );
   scalarSettings.setColorRampShader( fcn );
+  QgsMeshStrokeWidth edgeStrokeWidth;
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshlayer.cpp]: QgsMeshStrokeWidth -> QgsMeshStrokeWidthSettings (or QgsMeshVariableStrokeWidthSettings if it does not contain fixed width)

File: src/core/mesh/qgsmeshlayerrenderer.cpp [hunk 5]
```diff
@@ -456,29 +456,17 @@ void QgsMeshLayerRenderer::renderScalarDataset()
...
   const QVector<QgsMeshEdge> edges = mTriangularMesh.edges();
   const QVector<QgsMeshVertex> vertices = mTriangularMesh.vertices();
   const QList<int> egdesInExtent = mTriangularMesh.edgeIndexesForRectangle( context.mapExtent() );
   const QSet<int> nativeEdgesInExtent = QgsMeshUtils::nativeEdgesFromEdges( egdesInExtent,
                                         mTriangularMesh.edgesToNativeEdges() );
-  std::unique_ptr<QgsColorRampShader> shader( new QgsColorRampShader( scalarSettings.colorRampShader() ) );
-  QList<QgsColorRampShader::ColorRampItem> colorRampItemList = shader->colorRampItemList();
-  const QgsColorRampShader::Type classificationType = shader->colorRampType();
+
+  QgsMeshPlotter edgePlotter;
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshlayerrenderer.cpp]: QgsMeshPlotter -> QgsMeshEdgeRenderer

File: src/core/mesh/qgsmeshlayerrenderer.cpp [hunk 6]
```diff
@@ -456,29 +456,17 @@ void QgsMeshLayerRenderer::renderScalarDataset()
...
+  edgePlotter.setStrokeColor( QgsMeshStrokeColor( scalarSettings.colorRampShader() ) );
+  edgePlotter.setStrokeWidth( QgsMeshStrokeWidth( scalarSettings.edgeStrokeWidth() ) );
+  edgePlotter.setStrokeWidthUnit( scalarSettings.edgeStrokeWidthUnit() );
+  edgePlotter.setOpacity( scalarSettings.opacity() );
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshlayerrenderer.cpp:469]: maybe assign copy of scalarSettings to edgePlotter?

File: src/core/mesh/qgsmeshrenderersettings.h [hunk 7]
```diff
@@ -524,6 +512,12 @@ class CORE_EXPORT QgsMeshRendererVectorSettings
      */
     void setColorRampShader( const QgsColorRampShader &colorRampShader );
 
+    /**
+     * Returns the stroke coloring used to render vector datasets
+     * \since QGIS 3.14
+     */
+    QgsMeshStrokeColor vectorStrokeColoring() const;
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshrenderersettings.h]: vectorStrokeColor

File: src/core/mesh/qgsmeshplotter.h [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+#ifndef SIP_RUN
+
+/**
+ * \ingroup core
+ *
+ * Class plotting mesh dataset with stroke with and color varying dependending on dataset value.
+ *
+ * \since QGIS 3.14
+ */
+class CORE_EXPORT QgsMeshPlotter
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshplotter.h]: Do we need to export it outside qgis_core? If used internally in mesh renderer I would keep it private and do not export it

File: src/core/mesh/qgsmeshplotter.h [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+};
+
+/**
+ * \ingroup core
+ *
+ * Class defining stroke width to render dataset. The width can vary depending on dataset value.
+ *
+ * \since QGIS 3.14
+ */
+class CORE_EXPORT QgsMeshStrokeWidth
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshplotter.h]: QgsMeshStrokeWidth -> QgsMeshStrokeWidthSettings and move to the file with other mesh settings

for any calculation/logic I would suggest to move it to the mesh stroke/edge/1d renderer

File: src/core/mesh/qgsmeshplotter.h [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+    //! Returns whether the stroke width is varying
+    void setIsWidthVarying( bool isWidthVarying );
+
+    //! Returns the fixed stroke width
+    double fixedStrokeWidth() const;
+    //! Sets the fixed stroke width
+    void setFixedStrokeWidth( double fixedWidth );
+
+    //! Returns the varying stroke width depending on value, if not varying returns the fixed stroke width
+    double strokeWidth( double value ) const;
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshplotter.h]: I would maybe keep this QgsMeshStrokeWidth class just as a settings container and move this logic mesh edge rendererer

File: src/core/mesh/qgsmeshplotter.cpp [hunk 11]
```diff
@@ -0,0 +1,682 @@
...
+  if ( itemList.count() == 1 )
+  {
+    breakColors.append( itemList.first().color );
+    return;
+  }
+
+  double lastValue = itemList.at( itemList.count() - 2 ).value;
+
+
+  if ( value2 <= itemList.first().value ) // completly out of range and lesser
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshplotter.cpp]: completly -> completely

Comment (PeterPetrik) [reply src/core/mesh/qgsmeshplotter.cpp]: lesser -> less

File: src/core/mesh/qgsmeshplotter.cpp [hunk 12]
```diff
@@ -0,0 +1,682 @@
...
+  {
+    breakColors.append( itemList.first().color ); // render only the first color in the whole range
+    return;
+  }
+
+  if ( value1 > lastValue ) // completly out of range and greater
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshplotter.cpp]: completly -> completely

File: src/core/mesh/qgsmeshplotter.cpp [hunk 13]
```diff
@@ -0,0 +1,682 @@
...
+      breakValues.append( itemList.last().value );
+    else
+      breakValues.append( value2 );
+  }
+  breakColors.append( color2 );
+  gradients.append( makeSimpleLinearGradient( color1, color2 ) );
+
+
+
+  return;
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshplotter.cpp]: remove

File: src/core/mesh/qgsmeshplotter.cpp [hunk 14]
```diff
@@ -0,0 +1,682 @@
...
+  if ( itemList.empty() )
+    return;
+
+  if ( itemList.count() == 1 )
+  {
+    breakColors.append( itemList.first().color );
+    return;
+  }
+
+  if ( value2 <= itemList.first().value ) // completly out of range and lesser
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshplotter.cpp]: completly -> completely , lesser -> less

**Review by PeterPetrik — COMMENTED** [review]
In general I am very happy with the implementation and GUI and the code. I have only suggestion to shuffle the new classes a bit to follow the logic elsewhere in the code

1. for storing settings, create a new "Settings" class with only plain setters/getters (for color and width)
2. for rendering, create a new "Renderer" class where you put a logic how to do something with the settings and render the line (instead of QgsMeshPlotter and logic-part of QgsMeshStroke&Color)

**Comment by PeterPetrik** [comment]
we discussed and maybe instaed of Renderer, we can call it QgsMeshLineSymbol (similar to QgsVectorSymbol)

**Comment by PeterPetrik** [comment]
ok after discussion with Martin:

1. QgsMeshPlotter rename to QgsInterpolatedLineRenderer. And move outside the core/mesh folder to core or code/symbology
2. QgsMeshStrokeWidth rename to QgsInterpolatedLineWidth and keep the current implementation (no code splitting)
3. detto for stroke color

**Comment by wonder-sk** [comment]
the reason for the above suggestions is that this will be a very welcome addition to vector layer symbology as well, so best to keep if we do not make it mesh-specific. I am sure sooner or later someone will create a QgsSymbolLayer implementation making use of this code :-)

**Pull request closed without merging**